src/flags.rs

```diff
@@ -0,0 +1,478 @@
+                self.profile_generate = Some(value.expect("-Cprofile-generate must have a value"));
+            }
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#profile-use
+            "-Cprofile-use" => {
+                self.profile_use = Some(value.expect("-Cprofile-use must have a value"));
+            }
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#control-flow-guard
+            "-Ccontrol-flow-guard" => self.control_flow_guard = value.or(Some("true".into())),
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#lto
+            "-Clto" => self.lto = value.or(Some("true".into())),
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#relocation-model
+            "-Crelocation-model" => {
+                self.relocation_model = Some(value.expect("-Crelocation-model must have a value"))
+            }
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#embed-bitcode
+            "-Cembed-bitcode" => {
+                self.embed_bitcode = value.map_or(Some(true), |val| arg_to_bool(val.as_ref()));
+            }
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#force-frame-pointers
+            "-Cforce-frame-pointers" => {
+                self.force_frame_pointers =
+                    value.map_or(Some(true), |val| arg_to_bool(val.as_ref()));
+            }
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#link-dead-code
+            "-Clink-dead-code" => {
+                self.link_dead_code = value.map_or(Some(true), |val| arg_to_bool(val.as_ref()))
+            }
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#no-redzone
+            "-Cno-redzone" => {
+                self.no_redzone = value.map_or(Some(true), |val| arg_to_bool(val.as_ref()));
+            }
+            // https://doc.rust-lang.org/rustc/codegen-options/index.html#soft-float
+            // Note: This flag is now deprecated in rustc.
+            "-Csoft-float" => {
+                self.soft_float = value.map_or(Some(true), |val| arg_to_bool(val.as_ref()));
+            }
+            // https://doc.rust-lang.org/beta/unstable-book/compiler-flags/branch-protection.html
+            // FIXME: Drop the -Z variant and update the doc link once the option is stabilised
+            "-Zbranch-protection" | "-Cbranch-protection" => {
+                self.branch_protection =
+                    Some(value.expect("-Zbranch-protection must have a value"));
+            }
+            _ => {}
+        }
+    }
+
+    // Rust and clang/cc don't agree on what equivalent flags should look like.
+    pub fn cc_flags(
+        &self,
+        build: &Build,
+        path: &PathBuf,
+        family: &ToolFamily,
+        target: &Target,
+    ) -> Vec<OsString> {
+        let push_if_supported = |flags: &mut Vec<OsString>, flag: OsString| {
```

Comment:
```suggestion
        let push_if_supported = |flags: &mut Vec<OsString>, flag: String| {
```


Given that all usages of this function creates `String`, it would be more ergonomic to pass it to the function than doing additional `.into()`

Reply:
The into() will be done anyway because it needs to be an OsString for `is_flag_supported` and to actually push it to the vector. Should I just move the into()s inside the helper function then?

Reply:
It'd have to be something along the lines of
```
        let mut push_if_supported = |flag: String| {
            let flag = OsString::from(flag);
            if build
                .is_flag_supported_inner(&flag, path, target)
                .unwrap_or(false)
            {
                flags.push(flag);
            } else {
                build.cargo_output.print_warning(&format!(
                    "Inherited flag {:?} is not supported by the currently used CC",
                    flag
                ));
            }
        };
```

Reply:
Some of the flags are &str as well so if we start taking String it'll be &str -> String -> OsString for those instead of &str -> OsString directly as it is now. For the ones that use format!() and do create a String it'd just be moving the call converting them to OsString. I can do it if you'd like me to but I think this change adds some overhead for certain flags.

Reply:
Maybe we can use `impl Into<OsString>` instead?

That'd work with `&str`, `String` or `OsString`, and no extra allocation.

Reply:
```
rustc E0562: `impl Trait` is now allowed in closure parameters
```
Well that's annoying. If we make `push_if_supported` into a normal function to get around this then we'd need to explicitly pass everything it needs and the calls would end up looking like
```
push_if_supported("-fno-vectorize", flags, build, path, target)
```
Which is even less ergonomic than the `.into()s` we have there at the moment.

Reply:
E0562 seems to be about returning impl Traits from closure?

https://doc.rust-lang.org/error_codes/E0562.html

I was thinking about

```rust
        let mut push_if_supported = |flag: impl Into<OsString>| {
            let flag = OsString::from(flag);
            if build
                .is_flag_supported_inner(&flag, path, target)
                .unwrap_or(false)
            {
                flags.push(flag);
            } else {
                build.cargo_output.print_warning(&format!(
                    "Inherited flag {:?} is not supported by the currently used CC",
                    flag
                ));
            }
        };
```

Reply:
Thanks, in that case maybe we should keep the existing code.

I know cursed way to work around this (create a struct and put the fn as associated fn), but it's simply not worth the effort here.

Reply:
yeah I agree - all the other options are unfortunately even more messy.